File: binary_search/bsearch.php
<?php

/**
 * complexity O(log(N))
 */
function binarySearch($n, $arr) {
    $lBound = 0;
    $hBound = count($arr) - 1;

    while(true) {
        $curIn = ceil(($lBound + $hBound) / 2);
        $curVal = $arr[$curIn];

        if($curVal === $n) return (int)$curIn;
        else if($lBound > $hBound) return count($arr);
        else {
            if($curVal < $n) $lBound = $curIn + 1;
            else $hBound = $curIn - 1;
        }
    }
}

assert(0 === binarySearch(1, [1, 5, 6, 9, 12]));
assert(4 === binarySearch(12, [1, 5, 6, 9, 12]));
assert(2 === binarySearch(6, [1, 5, 6, 9, 12]));
assert(5 === binarySearch(7, [1, 5, 6, 9, 12]));

File: binary_search/bsearch.clj
(defn bsearch-helper [xs x lBound hBound]
  (if (> lBound hBound) (count xs)
  (let [curIn (int (/ (+ lBound hBound) 2)) curVal (nth xs curIn)]
    (cond
      (= curVal x) curIn
      (> curVal x) (recur xs x lBound (- curIn 1))
      :else (recur xs x (+ curIn 1) hBound)
    ))))

(defn binary-search [xs x]
  (bsearch-helper xs x 0 (- (count xs) 1)))


(assert (= (binary-search '(1, 5, 6, 9, 12) 1) 0)
(assert (= (binary-search '(1, 5, 6, 9, 12) 12) 4)
(assert (= (binary-search '(1, 5, 6, 9, 12) 6) 2)
(assert (= (binary-search '(1, 5, 6, 9, 12) 7) 5)
